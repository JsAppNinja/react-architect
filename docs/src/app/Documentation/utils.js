export const splitOnCaps = s => s.split(/(?=[A-Z1-9])/);

function getString(s) {
  return !(s instanceof Array) ? splitOnCaps(s) : s;
}

export const toDashedName = (s) => getString(s).map(s => s.toLowerCase()).join('-');
export const toTitle = (s) => getString(s).join(' ');
