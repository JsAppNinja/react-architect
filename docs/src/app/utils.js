import React from 'react';
export const githubHref = 'https://github.com/mlaursen/react-md';

const lorems = [
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus at placerat nulla. Morbi eu erat diam. Donec malesuada mauris in neque efficitur luctus. Vestibulum porttitor elit velit, pellentesque imperdiet tellus maximus in. Phasellus consectetur lacus nunc, quis efficitur felis feugiat ac. Vivamus non erat laoreet, volutpat tellus sit amet, porttitor orci. Sed scelerisque id arcu at ornare. Suspendisse potenti. Phasellus eu posuere mi. Fusce venenatis odio dui, ac accumsan lacus dapibus et. Praesent euismod nulla at porta elementum. Nunc ullamcorper velit aliquam velit faucibus tempus.',
  'Duis ligula elit, elementum a nisl a, venenatis fermentum libero. Cras bibendum dignissim nisl, sed sodales tellus scelerisque sed. Sed at enim semper, imperdiet elit et, venenatis magna. Morbi vestibulum purus in ante placerat viverra. Nullam molestie arcu ut dolor rutrum varius. Sed vehicula laoreet volutpat. Quisque feugiat, nisl in sollicitudin consectetur, mi metus aliquet urna, ac sagittis nulla nulla in diam. Fusce id enim auctor ante fringilla auctor in et quam.',
  'Donec pulvinar ligula augue, et tincidunt velit bibendum non. Aenean id metus nunc. Donec non feugiat nisl. In accumsan ligula purus, ut vulputate elit tempus sit amet. Nullam volutpat semper lectus. Nunc a libero tellus. Duis posuere nisi diam, nec euismod risus interdum sit amet. Pellentesque at lacus bibendum, cursus ante non, maximus arcu. Morbi id leo nulla. Integer quis feugiat quam. Etiam auctor pharetra vestibulum.',
  'Nulla sodales metus metus. Sed a metus quis ipsum sagittis placerat. Nunc sit amet molestie purus, nec rutrum quam. Praesent elementum risus eget mattis condimentum. Morbi dignissim, dolor et finibus sollicitudin, nulla enim molestie est, eget tincidunt magna libero et dolor. Fusce tempor sapien et urna sodales, at laoreet nunc egestas. Aenean eget arcu in felis ultrices vehicula. Maecenas interdum viverra sem vitae sodales. Duis tincidunt varius nunc sed viverra. Duis sit amet semper ex, ultrices aliquet mi. Donec a congue sem, ac facilisis ex.',
  'Proin venenatis tortor et ex viverra, ac facilisis mi ullamcorper. Aenean eget est bibendum, fermentum enim eget, semper orci. Nunc magna mauris, fermentum sit amet diam ut, semper malesuada justo. Fusce faucibus tellus eget arcu imperdiet, aliquet placerat massa molestie. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Maecenas gravida quam id ultricies accumsan. Sed porta accumsan libero sed venenatis. Morbi scelerisque, ex in gravida porttitor, ipsum erat sodales urna, mattis mattis nibh ligula quis tellus. Pellentesque consequat et metus scelerisque dapibus. Praesent at maximus ligula. Phasellus vel orci tortor.',
  'Mauris posuere nisl sit amet justo condimentum volutpat. Aliquam venenatis rhoncus est, tristique semper tortor iaculis et. Donec pretium metus nec sodales mollis. Ut vel ipsum sapien. Fusce sit amet libero efficitur orci aliquet scelerisque a ut arcu. In hac habitasse platea dictumst. Sed turpis libero, dictum nec imperdiet nec, maximus eget urna. Sed vel lorem sit amet nisi vulputate mattis. In in odio erat. Aenean euismod lorem nec viverra auctor. Mauris id ipsum maximus, tempus dui nec, mollis felis. Etiam sodales metus sit amet lacus aliquet tristique. Nulla in lacus massa. Morbi quis lorem vitae dolor maximus luctus vel vel eros. Phasellus posuere in purus eget convallis.',
  'Aliquam mauris justo, ornare quis quam fermentum, mollis volutpat dolor. Donec sagittis velit non nisl porttitor, id lacinia tortor vulputate. Suspendisse posuere at felis eu fermentum. Donec at vehicula leo, id sollicitudin ipsum. Donec dapibus sodales erat at malesuada. Mauris sollicitudin velit sed arcu molestie congue. Suspendisse ac arcu enim. Sed ac urna id mi rutrum ornare sit amet in nisl. Phasellus at tempus magna. Nulla facilisi. Mauris pharetra finibus mi ut facilisis. Duis mattis odio id quam vestibulum, id consequat ex fringilla. Aenean sollicitudin accumsan sollicitudin. Duis euismod maximus tempor. Phasellus mi enim, bibendum at nisi vel, cursus venenatis sem. Praesent sagittis, nunc at rhoncus maximus, ligula turpis imperdiet eros, pharetra tincidunt tortor magna id sem.',
  'Sed eleifend sapien condimentum iaculis cursus. In vel pretium turpis. Duis molestie euismod turpis id consectetur. Nunc pharetra sapien libero, a volutpat erat finibus eget. Duis euismod eleifend felis consequat fringilla. Sed varius diam ligula, quis rutrum ex mollis non. Phasellus ac malesuada nibh. Praesent sit amet auctor urna. Quisque eget magna id ex dapibus vulputate eget vel magna.',
  'Ut viverra tempor placerat. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras neque diam, auctor eget sapien et, tincidunt blandit turpis. Nam non volutpat velit. Aliquam iaculis euismod nisl non facilisis. Ut blandit dignissim turpis in finibus. Cras neque augue, tempus non neque eu, ornare ultrices sem. Nullam ultricies mi quis semper pellentesque. Mauris mauris lacus, ullamcorper a iaculis quis, tempus id tellus. Phasellus in orci orci. Fusce consequat suscipit ipsum ac bibendum. Maecenas et leo eu leo iaculis varius. Nam consectetur laoreet nisl blandit condimentum.',
  'Ut dictum, sapien vel malesuada posuere, tortor tellus blandit risus, eget hendrerit quam diam eu nibh. Aliquam erat volutpat. Ut rhoncus augue ac nunc porttitor hendrerit. Pellentesque enim risus, auctor quis turpis eget, pulvinar malesuada urna. Nullam tortor mi, elementum id sodales eget, aliquam et nulla. Pellentesque efficitur mi eros, in rutrum metus ultrices vel. Proin vitae diam sit amet nunc pretium malesuada. Fusce at nisi scelerisque, ultrices tortor nec, finibus velit. Donec molestie augue eget leo rutrum, nec ornare arcu gravida. Quisque tincidunt risus elit, ut condimentum eros egestas congue.',
];
export function loremIpsum(paras = 3) {
  let ipsums = [];
  for(let i = 0; i < paras; i++) {
    ipsums.push((
      <p key={i}>{lorems[Math.floor(Math.random() * lorems.length)]}</p>
    ));
  }
  return ipsums;
}

export const numstr = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten'];

export const randomImage = ({ width, height, time } = {}) => {
  width = !width ? 200 : width;
  height = !height ? 200 : height;
  time = typeof time === 'undefined' ? Date.now() : time;
  return `https://unsplash.it/${width}/${height}/?random&time=${time}`;
};

export const hostPrefix = ''; // change to '/react-md' for production
export const imgPrefix = hostPrefix + '/imgs';
