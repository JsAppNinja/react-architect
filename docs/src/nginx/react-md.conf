server {
  # Redirect http requests to https
  listen       80;
  server_name  react-md.mlaursen.com;

  return 301 https://$server_name$request_uri;
}

server {
  listen       443 ssl;
  server_name  react-md.mlaursen.com;

  ssl_protocols               TLSv1 TLSv1.1 TLSv1.2;
  ssl_certificate             /etc/letsencrypt/live/react-md.mlaursen.com/fullchain.pem;
  ssl_certificate_key         /etc/letsencrypt/live/react-md.mlaursen.com/privkey.pem;
  ssl_prefer_server_ciphers   on;
  ssl_ciphers                 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
  ssl_dhparam                 /etc/ssl/certs/dhparam.pem;


  # Enable nginx static asset caching and serving since it is faster than express
  # This will server all the files in the public/ folder EXCEPT for the custom themes
  location ~ /(sassdoc$|sassdoc\/|favicon.ico|react-md.png|robots.txt|manifest.json|(?!themes\/.*)\.css$|.+\.(js|map|svg|png)$) {
    root   /var/www/react-md/v1.1.x/docs/public;

    proxy_set_header       Host $host;
    proxy_cache            STATIC;
    proxy_cache_valid      200  1d;
    proxy_cache_use_stale  error timeout invalid_header updating
                           http_500 http_502 http_503 http_504;
  }

  
  # This doesn't seem to be working right now. Need to debug a bit more
  # location /api {
  #   # Change to static IP
  #   allow http://localhost;
  #   deny all;
  # }

  location / {
    proxy_pass          http://localhost:8080;
    proxy_http_version  1.1;
    proxy_set_header    Upgrade $http_upgrade;
    proxy_set_header    Connection 'upgrade';
    proxy_set_header    Host $host;
    proxy_cache_bypass  $http_upgrade;
  }
}
