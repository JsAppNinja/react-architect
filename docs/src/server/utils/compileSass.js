const path = require('path');
const Promise = require('bluebird');
const nodeSass = require('node-sass');

const sass = Promise.promisify(nodeSass.render);

const { ACCENTABLE_COLORS, PRIMARY_COLORS, SECONDARY_HUES } = require('../../shared/constants/colors');

const DARK_THEME_STYLES = `
.example-code,
.code-block {
  background: $md-grey-300;
}

a:not(.md-btn) {
  color: $md-purple-a-400;

  &:visited {
    color: $md-purple-a-100;
  }
}
`;

function toColorName(color) {
  return color && color.replace('_', '-');
}

function validatePrimary(color) {
  return color && PRIMARY_COLORS.indexOf(color) !== -1;
}

function validateSecondary(color, hue) {
  return color && hue && ACCENTABLE_COLORS.indexOf(color) !== -1 && SECONDARY_HUES.indexOf(hue) !== -1;
}

/**
 * Attempts to compiles sass based off of a css file name. The filename should be formatted
 * as 'PRIMARY_COLOR-SECONDARY_COLOR-HUE-dark?.css' where `dark` is optional.
 *
 * @param {String} name - The CSS filename to convert into styles.
 * @return {Promise} a promise with the buffer of compiles styles
 */
module.exports = function compileSass(name) {
  const [p, s, hue, dark] = name.split('-');
  const primary = toColorName(p);
  const secondary = toColorName(s);

  if (!validatePrimary(primary)) {
    return Promise.reject(new Error(`Invalid primary color: ${primary}`));
  } else if (!validateSecondary(secondary, parseInt(hue, 10))) {
    return Promise.reject(new Error(`Invalid secondary color: ${secondary}-${hue}`));
  }

  const primaryColor = `$md-${primary}-500`;
  const secondaryColor = `$md-${secondary}-a-${hue}`;
  const darkThemeStyles = dark ? DARK_THEME_STYLES : '';

  const styles = `
@import 'react-md';

$md-primary-color: $md-light-blue-500;
$md-secondary-color: $md-deep-orange-a-200;
@include react-md-theme-everything(${primaryColor}, ${secondaryColor}, ${!dark}, 'custom-theme');

.custom-theme {
  .banner {
    background: ${primaryColor};
  }

  .react-md-logo {
    text,
    #Group > path:first-child,
    path#Oval-210 {
      fill: ${secondaryColor};
    }
  }

  ${darkThemeStyles}
}

@media #{$md-tablet-media} {
  .custom-theme .phone-size-container .md-toolbar:not(.md-toolbar--inset)::before {
    background: ${primaryColor};
  }
}
  `;

  return sass({
    data: styles,
    includePaths: [path.resolve(process.cwd(), '..', 'src', 'scss')],
    outputStyle: 'compressed',
  }).then((stats, error) => {
    if (error) {
      return Promise.reject(error);
    }

    return stats.css;
  });
};
