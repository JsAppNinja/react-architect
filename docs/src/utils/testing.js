/* eslint-env jest */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { StaticRouter } from 'react-router';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { render, shallow, mount } from 'enzyme';
import rootReducer from 'state';

/**
 * I'm lazy.
 *
 * Could be lazier and do `expectSnapshot` though..
 */
export function createSnapshot(children) {
  return renderer.create(children).toJSON();
}

/**
 * A simple test wrapper to create a snapshot with react-test-renderer and a component that
 * uses something from react-router.
 */
export function createRouterSnapshot(children, location = '/', context = {}) {
  return createSnapshot(
    <StaticRouter location={location} context={context}>
      {children}
    </StaticRouter>
  );
}

/**
 * A simple test wrapper to create a snapshot with react-test-renderer and a component that
 * requires the redux provider.
 *
 * > It is better to test "pure" components instead of these, so this probably shouldn't
 * > get used much.
 */
export function createReduxSnapshot(children, state) {
  const store = createStore(rootReducer, state);
  return createSnapshot(
    <Provider store={store}>
      {children}
    </Provider>
  );
}

/**
 * A simple test wrapper to create a snapshot with react-test-renderer and a component that
 * requires both the redux Provider and something from react-router.
 *
 * > It is better to test "pure" components instead of these, so this probably shouldn't
 * > get used much.
 */
export function createReduxRouterSnapshot(children, state, location = '/', context = {}) {
  const store = createStore(rootReducer, state);
  return createSnapshot(
    <Provider store={store}>
      <StaticRouter location={location} context={context}>
        {children}
      </StaticRouter>
    </Provider>
  );
}

/**
 * This is a simple test wrapper to create a snapshot with enzyme's render instead of react-test-renderer
 * and when the component or child component relies on react-router. This should be used when a component
 * or child component uses findDOMNode or the CSSTransitionGroup, otherwise the createRouterSnapshot should
 * be used.
 */
export function renderRouterSnapshot(children, location = '/', context = {}) {
  return render(
    <StaticRouter location={location} context={context}>
      {children}
    </StaticRouter>
  );
}

/**
 * This is a simple test wrapper to create a snapshot with enzyme's render instead of react-test-renderer
 * and when the component or child component requires the redux Provider. This should be used when a component
 * or child component uses findDOMNode or the CSSTransitionGroup, otherwise the createReduxSnapshot should
 * be used.
 */
export function renderReduxSnapshot(children, state) {
  const store = createStore(rootReducer, state);
  return render(
    <Provider store={store}>
      {children}
    </Provider>
  );
}

/**
 * This is a simple test wrapper to create a snapshot with enzyme's render instead of react-test-renderer
 * and when the component or child component requires both the redux Provider and something from react-router.
 * This should be used when a component * or child component uses findDOMNode or the CSSTransitionGroup,
 * otherwise the createReduxRouterSnapshot should be used.
 */
export function renderReduxRouterSnapshot(children, state, location = '/', context = {}) {
  const store = createStore(rootReducer, state);
  return render(
    <Provider store={store}>
      <StaticRouter location={location} context={context}>
        {children}
      </StaticRouter>
    </Provider>
  );
}

/**
 * A simple wrapper to use enzyme's shallow rendering when the component or a child
 * requires something from react-router.
 */
export function shallowWithRouter(children, location = '/', context = {}) {
  return shallow(
    <StaticRouter location={location} context={context}>
      {children}
    </StaticRouter>
  );
}


/**
 * A simple wrapper to use enzyme's mount rendering when the component or a child
 * requires something from react-router.
 */
export function mountWithRouter(children, location = '/', context = {}) {
  return mount(
    <StaticRouter location={location} context={context}>
      {children}
    </StaticRouter>
  );
}

/**
 * A simple wrapper to use enzyme's shallow rendering when the component or a child
 * requires something from react-redux.
 *
 * > It is better to test "pure" components instead of these, so this probably shouldn't
 * > get used much.
 */
export function shallowWithProvider(children, state) {
  const store = createStore(rootReducer, state);
  return shallow(<Provider store={store}>{children}</Provider>);
}

/**
 * A simple wrapper to use enzyme's mount rendering when the component or a child
 * requires something from react-redux.
 *
 * > It is better to test "pure" components instead of these, so this probably shouldn't
 * > get used much.
 */
export function mountWithProvider(children, state) {
  const store = createStore(rootReducer, state);
  return mount(<Provider store={store}>{children}</Provider>);
}

/**
 * A utility function that will override the console's behavior and throw errors
 * if console.error occurs. This is useful when you want to consider React warnings
 * as failures.
 */
export function captureConsole(match = null) {
  const console = global.console;

  beforeAll(() => {
    global.console = {
      ...console,
      error: (error) => {
        if (match === null || error.match(match)) {
          throw new Error(error);
        }
      },
    };
  });

  afterAll(() => {
    global.console = console;
  });
}
