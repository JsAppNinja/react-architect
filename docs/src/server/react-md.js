import React from 'react';
import { renderToString } from 'react-dom/server';
import { match } from 'react-router';

import { getPageTitle } from 'utils/StringUtils';
import { getDrawerType } from 'utils/MediaUtils';
import Root from './Root';
import routes from 'routes';
import configureStore from 'stores/configureStore';

export default function reactMD(req, res) {
  const userAgent = req.header('user-agent');
  const mobile = !!userAgent.match(/mobile/i);
  const tablet = !!userAgent.match(/ipad/i);

  const toolbarTitle = getPageTitle(req.url);
  const initialState = {
    ui: {
      drawer: {
        initiallyOpen: !mobile,
        initialDrawerType: getDrawerType(mobile, tablet),
        inactive: !toolbarTitle,
        toolbarTitle,
      },
      media: {
        mobile,
        tablet,
        desktop: !mobile && !tablet,
      },
    },
  };

  const store = configureStore(initialState);

  match({ routes, location: req.url }, (error, redirectLocation, props) => {
    if (error) {
      req.status(500).send(error.message);
    } else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search);
    } else if (props) {
      res.render('index', {
        initialState: JSON.stringify(initialState),
        html: renderToString(<Root store={store} {...props} />),
      });
    } else {
      res.status(404).send('Not Found');
    }
  });
}
