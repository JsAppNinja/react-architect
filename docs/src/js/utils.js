import React from 'react';

import { FontIcon, Avatar } from 'react-md';

export const githubHref = 'https://github.com/mlaursen/react-md';

export const mainLinks = [{
  link: '',
  primaryText: 'Home',
  leftIcon: <FontIcon>home</FontIcon>,
}, {
  link: 'sassdoc',
  primaryText: 'SASS Doc',
  leftAvatar: <Avatar src="/imgs/sass-icon.png" alt="SASS Icon" />,
}];
