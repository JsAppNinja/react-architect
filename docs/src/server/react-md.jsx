import React from 'react';
import { renderToString } from 'react-dom/server';
import { match } from 'react-router';

import { getPageTitle } from 'utils/StringUtils';
import Root from './Root';
import routes from 'routes';
import configureStore from 'stores/configureStore';

function getDrawerType(mobile, ipad) {
  if (mobile) {
    return 'mobile';
  } else if (ipad) {
    return 'ipad';
  } else {
    return 'desktop';
  }
}

export default function reactMD(req, res) {
  const userAgent = req.header('user-agent');
  const mobile = !!userAgent.match(/mobile/i);
  const ipad = !!userAgent.match(/ipad/i);

  const toolbarTitle = getPageTitle(req.url);
  const initialState = {
    ui: {
      drawer: {
        initiallyOpen: !mobile,
        initialDrawerType: getDrawerType(mobile, ipad),
        inactive: !toolbarTitle,
        toolbarTitle,
      },
    },
  };

  const store = configureStore(initialState);

  match({ routes, location: req.url }, (error, redirectLocation, props) => {
    if (error) {
      req.status(500).send(error.message);
    } else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search);
    } else if (props) {
      res.render('index', {
        initialState: JSON.stringify(initialState),
        html: renderToString(<Root store={store} {...props} />),
      });
    } else {
      res.status(404).send('Not Found');
    }
  });
}
