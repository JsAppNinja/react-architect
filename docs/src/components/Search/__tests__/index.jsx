/* eslint-disable react/prop-types */
/* eslint-env jest */
import React from 'react';
import { mount, shallow } from 'enzyme';
import Waypoint from 'react-waypoint';
import { Button } from 'react-md';

import { renderRouterSnapshot, mountWithRouter, captureConsole } from 'utils/testing';
import { PureSearch as Search } from '../';

// Don't know how to surround with StaticRouter and then set this component's state
jest.mock('react-router-dom', () => ({ children, className, to }) => <a className={className} href={to}>{children}</a>);

const PROPS = {
  mobile: false,
  searching: true,
  search: () => {},
  searchNext: () => {},
  showSearch: () => {},
  hideSearch: () => {},
};

const RESULTS = [{
  name: 'Autocompletes',
  type: 'Info',
  ref: '/components/autocompletes',
}, {
  name: 'Autocompletes',
  type: 'Prop Types',
  ref: '/components/autocompletes?tab=1',
}];

describe('Search', () => {
  captureConsole();

  it('should render correctly while not searching', () => {
    const tree = renderRouterSnapshot(<Search {...PROPS} searching={false} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly while searching', () => {
    const tree = renderRouterSnapshot(<Search {...PROPS} searching />);
    expect(tree).toMatchSnapshot();
  });

  it('should call the showSearch prop when the icon is clicked and it is currently not searching', () => {
    const showSearch = jest.fn();
    const props = { ...PROPS, showSearch, searching: false };

    const search = mount(<Search {...props} />);
    let icon = search.find('.md-icon');
    expect(icon.length).toBe(1);
    icon.simulate('click');

    expect(showSearch).toBeCalled();

    showSearch.mockClear();
    search.setProps({ searching: true });
    icon = search.find('.md-icon');
    expect(icon.length).toBe(2);
    icon.first().simulate('click');
    icon.at(1).simulate('click');

    expect(showSearch).not.toBeCalled();
  });

  it('should create an animating timeout when the searching prop changes to false', () => {
    const search = shallow(<Search {...PROPS} searching={false} />);
    expect(search.state('animating')).toBe(false);

    search.setProps({ searching: true });
    expect(setTimeout).not.toBeCalled();

    search.setProps({ searching: false });
    expect(setTimeout).toBeCalled();
    expect(search.state('animating')).toBe(true);
    jest.runOnlyPendingTimers();
    expect(search.state('animating')).toBe(false);
  });

  it('should clear any timeouts if it unmounts', () => {
    jest.useFakeTimers();
    const search = shallow(<Search {...PROPS} />);
    expect(setTimeout).not.toBeCalled();
    search.setProps({ searching: false });
    expect(setTimeout).toBeCalled();

    search.unmount();
    jest.runAllTimers();
  });

  it('should set the internal value state when the value changes', () => {
    const search = mount(<Search {...PROPS} />);
    expect(search.state('value')).toBe('');
    const input = search.find('input');
    expect(input.length).toBe(1);

    const value = 'data tab';
    input.simulate('change', { target: { value } });
    expect(search.state('value')).toBe(value);
  });

  it('should clear the internal value state when the searching prop changes', () => {
    const search = mount(<Search {...PROPS} />);
    search.setState({ value: 'buttons' });

    search.setProps({ searching: false });
    expect(search.state('value')).toBe('');
  });

  it('should call the search prop when the value changes in the autocomplete', () => {
    const searchFn = jest.fn();
    const search = mount(<Search {...PROPS} search={searchFn} searching />);
    const input = search.find('input');
    expect(input.length).toBe(1);

    const value = 'data tab';
    input.simulate('change', { target: { value } });
    expect(searchFn).toBeCalledWith(value);
  });

  it('should update the data list when the results prop changes', () => {
    const search = shallow(<Search {...PROPS} />);
    expect(search.state('data')).toEqual([]);

    const instance = search.instance();
    const mapToLink = instance.mapToLink;
    // Mock it just to make sure it is called
    instance.mapToLink = jest.fn(item => mapToLink(item));

    search.setProps({ results: RESULTS });

    expect(instance.mapToLink).toBeCalled();
    expect(instance.mapToLink.mock.calls.length).toBe(RESULTS.length);
  });

  it('should inject the Waypoint component when there are more search results', () => {
    let search = mountWithRouter(<Search {...PROPS} />);
    let input = search.find('input');
    input.simulate('focus');
    input.simulate('change', { target: { value: 'aut' } });
    expect(search.find(Waypoint).length).toBe(0);
    const props = {
      ...PROPS,
      results: RESULTS,
      meta: {
        next: '/some-url',
        previous: null,
        start: 0,
        limit: 2,
        total: 300,
      },
    };

    search = mountWithRouter(<Search {...props} />);
    input = search.find('input');
    input.simulate('focus');
    input.simulate('change', { target: { value: 'aut' } });
    expect(search.find(Waypoint).length).toBe(1);
  });

  it('should call the hideSearch prop when the hide button is closed', () => {
    const hideSearch = jest.fn();
    const search = mount(<Search {...PROPS} hideSearch={hideSearch} />);
    expect(hideSearch).not.toBeCalled();

    const btn = search.find(Button);
    expect(btn.length).toBe(1);

    btn.simulate('click');
    expect(hideSearch).toBeCalled();
  });
});
