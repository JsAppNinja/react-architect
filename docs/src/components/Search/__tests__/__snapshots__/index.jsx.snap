// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly while not searching 1`] = `
<div
  class="search md-grid md-grid--no-spacing"
>
  <div
    aria-expanded="false"
    aria-haspopup="true"
    class="md-layover md-full-width md-menu-container md-autocomplete-container search__autocomplete"
    id="documentation-search-menu"
  >
    <span
      aria-hidden="true"
    />
    <div
      class="md-text-field-container md-full-width md-text-field-container--input md-autocomplete"
    >
      <div
        class="md-text-field-icon-container"
      >
        <i
          class="md-icon material-icons md-text-field-icon md-text-field-icon--positioned"
        >
          search
        </i>
        <div
          class="md-text-field-divider-container md-text-field-divider-container--grow"
        >
          <input
            aria-autocomplete="list"
            autocomplete="off"
            class="md-text-field md-text-field--margin md-full-width md-text search__input"
            id="documentation-search"
            placeholder="Search"
            type="text"
            value=""
          />
          <hr
            class="md-divider md-divider--text-field md-divider--expand-from-left"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search should render correctly while searching 1`] = `
<div
  class="search md-grid md-grid--no-spacing"
>
  <div
    aria-expanded="false"
    aria-haspopup="true"
    class="md-layover md-full-width md-menu-container md-autocomplete-container search__autocomplete"
    id="documentation-search-menu"
  >
    <span
      aria-hidden="true"
    />
    <div
      class="md-text-field-container md-full-width md-text-field-container--input md-autocomplete"
    >
      <div
        class="md-text-field-icon-container"
      >
        <i
          class="md-icon material-icons md-text-field-icon md-text-field-icon--positioned"
        >
          search
        </i>
        <div
          class="md-text-field-divider-container md-text-field-divider-container--grow"
        >
          <input
            aria-autocomplete="list"
            autocomplete="off"
            class="md-text-field md-text-field--margin md-full-width md-text search__input search__input--visible search__input--active"
            id="documentation-search"
            placeholder="Search"
            type="text"
            value=""
          />
          <hr
            class="md-divider md-divider--text-field md-divider--expand-from-left"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    class="md-btn md-btn--icon md-pointer--hover md-inline-block"
    id="documentation-search-hide"
    type="button"
  >
    <i
      class="md-icon material-icons md-text--inherit"
    >
      close
    </i>
  </button>
</div>
`;
